MapboxNavigationUI/NavigationMapView.swift

```diff
@@ -13,4 +17,77 @@ public class NavigationMapView: MGLMapView {
             super.locationManager(manager, didUpdateLocations: locations)
         }
     }
+    
+    public func annotate(_ route: Route) {
+        guard let style = style else {
+            return
+        }
+        
+        if let line = style.layer(withIdentifier: routeLayerIdentifier) {
+            style.removeLayer(line)
+        }
+        
+        if let lineCasing = style.layer(withIdentifier: routeLayerCasingIdentifier) {
+            style.removeLayer(lineCasing)
+        }
+        
+        if let source = style.source(withIdentifier: sourceIdentifier) {
+            style.removeSource(source)
+        }
+        
+        let polyline = shape(for: route)
+        
+        if let source = style.source(withIdentifier: sourceIdentifier) as? MGLShapeSource {
+            source.shape = polyline
+        } else {
+            let geoJSONSource = MGLShapeSource(identifier: sourceIdentifier, shape: polyline, options: nil)
+            style.addSource(geoJSONSource)
+            
+            let line = lineStyle(MGLLineStyleLayer(identifier: routeLayerIdentifier, source: geoJSONSource))
+            let lineCasing = lineCasingStyle(MGLLineStyleLayer(identifier: routeLayerCasingIdentifier, source: geoJSONSource))
+            
+            for layer in style.layers.reversed() {
+                if !(layer is MGLSymbolStyleLayer) &&
+                    layer.identifier != arrowLayerIdentifier && layer.identifier != arrowSourceIdentifier {
+                    style.insertLayer(line, above: layer)
+                    style.insertLayer(lineCasing, below: line)
+                    return
+                }
+            }
+        }
+    }
+    
+    public func shape(for route: Route) -> MGLShape? {
+        guard var coordinates = route.coordinates else {
+            return nil
+        }
+        
+        return MGLPolylineFeature(coordinates: &coordinates, count: route.coordinateCount)
+    }
+    
+    public func lineStyle(_ line: MGLLineStyleLayer) -> MGLLineStyleLayer {
```

Comment:
This method needs a more descriptive name. Imagine reading aloud this code:

```swift
let line = lineStyle(lineLayer)
```

“Let line be line style line layer”

The reason it’s difficult to come up with a descriptive name right now is that it needs to be scoped better. This method should be responsible for creating the layer too:

```swift
public func routeStyleLayer(identifier: String, source: MGLSource) -> MGLStyleLayer
```

Now the call site could look similar to this:

```swift
let line = routeStyleLayer(identifier: "routeLayerIdentifier", source: source)
```

“Let line be the route style layer with identifier …”